sources/osg/Program.js

```diff
@@ -263,105 +271,170 @@ utils.createPrototypeStateAttribute(
+                            errLink,
+                            "can't link program\nvertex shader:\n" +
+                                this._vertex.text +
+                                '\n fragment shader:\n' +
+                                this._fragment.text
+                        );
+
+                        if (this._onErrorToSpector(errLink)) return;
+                        this._logDebugShaders(gl, errLink);
+
+                        compileClean = false;
+                    } else {
+                        this._onCompilationToSpector(this.program);
                     }
+                }
+
+                this._bindProgramToSpector();
+
+                if (!compileClean) {
+                    // Any error, Any
+                    // Pink must die.
+                    this._activateFailSafe(gl);
+                }
+                this._uniformsCache = {};
+                this._attributesCache = {};
+
+                this.timeAndStart(' end link/check ');
+
+                this.cacheAttributeList(gl, window.Object.keys(this._attributeMap));
+                this.cacheUniformList(gl, window.Object.keys(this._uniformMap));
+
+                this.timeAndStart(' end link location ');
+                console.log('end gl COMPILATION ' + this._shaderName + ': ' + this._Accum + ' ms');
 
-                    this._bindProgramToSpector();
+                return compileClean;
+            },
 
-                    this._uniformsCache = {};
-                    this._attributesCache = {};
+            getAsyncCompiling: function() {
+                return this._asyncCompilation !== undefined;
+            },
 
-                    this.cacheUniformList(gl, uniformArray);
-                    this.cacheAttributeList(gl, attributeArray);
+            apply: function(state) {
+                if (this._nullProgram) return;
 
-                    this._dirty = false;
+                if (!this._gl) {
+                    this.setGraphicContext(state.getGraphicContext());
                 }
 
+                if (this._program && !this._dirty) {
+                    state.applyProgram(this._program);
+                    return;
+                }
+
+                var asyncShader = true;
+                if (!asyncShader) {
+                    this._compileClean = this.compile(state);
+                } else if (this._asyncCompilation === undefined) {
+                    // async shader compilation part
+                    this._compileClean = this.compile(state);
+                    // async compilation threshold ?
+                    // strlen of vertex/fragment
+                    // number of uniform
+                    this._asyncCompilation = Timer.instance().tick();
+                    return;
+                } else if (this._asyncCompilation) {
+                    // still waiting
+                    if (Timer.instance().tick() - this._asyncCompilation > 800) {
```

Comment:
Not a fan of hardcoded values. Hard to find later. Also in any case should be 800.0 I guess